File: js/completed/app_create.js
const algosdk = require("algosdk");
const fs = require("fs");
const validate = require("../validate");
const printError = require("../error-printer");
const { encodeAddress } = require("algosdk");

const challenge_id = "TBD"

const token = "";
const server = "https://testnet-api.algonode.cloud";
const port = 0;
const client = new algosdk.Algodv2(token, server, port);
const txids = [];

const secretKey =
  "sYgLa2BSnHCYG1tlugubFuoLYjGoPoHiM71JkONCn3zWYyP45GoVqoNmZ4b31uCqML4FzQ0uSEg4PXGhCNW9TA==";

// Decode the secretKey into a Uint8Array from base 64
// This will produce an array of length 64
const secret = new Uint8Array(Buffer.from(secretKey, "base64"));
const acct = {
  // The public key is the secret[32:], or the last 32 bytes
  // We encode it to the address which is easier to read and includes a checksum
  addr: encodeAddress(secret.slice(32)),
  // We need not do anything with the secret
  sk: secret,
};

const dir = "/home/ben/challenges"
// Read in the programs
const approval = fs.readFileSync(dir+"/approval.teal");
const clear = fs.readFileSync(dir+"/clear.teal");

const accessCode = new Uint8Array(Buffer.from("not-a-secret"));

(async function () {
  try {
    // Compile the teal programs to bytecode
    const approvalRes = await client.compile(approval).do()
    const approvalProgram = new Uint8Array(Buffer.from(approvalRes["result"], 'base64'));

    const clearRes = await client.compile(clear).do()
    const clearProgram = new Uint8Array(Buffer.from(clearRes["result"], 'base64'));

    // Get the suggested parameters from the Algod server. 
    // These include current fee levels and suggested first/last rounds.
    const createsp = await client.getTransactionParams().do();

    // Construct an asset create transaction. 
    // An asset create transaction is the same as an asset config transaction
    // with its asset id set to 0
    const txn = algosdk.makeApplicationCreateTxnFromObject({
      from: acct.addr,
      approvalProgram: approvalProgram,
      clearProgram: clearProgram,
      numGlobalByteSlices: 1,
      numLocalByteSlices: 0,
      numGlobalInts: 1,
      numLocalInts: 0,
      appArgs: [accessCode],
      suggestedParams: createsp,
    });

    // Sign the transaction. This should return a 
    // Uint8Array representing the bytes to be sent to the network
    const signed = txn.signTxn(acct.sk);

    // Send the transaction, returns the transaction id for 
    // the first transaction in the group
    const createDetails = await client.sendRawTransaction(signed).do();
    txids.push(createDetails.txId);

    // Wait for the transaction to be confirmed.
    const result = await algosdk.waitForConfirmation(client, createDetails.txId, 2);

    const app_id = result["application-index"]
    // Log out the confirmed round
    console.log("Confirmed round: " + result["confirmed-round"]);
    console.log("app id: " + app_id);

    // Call the app
    const callsp = await client.getTransactionParams().do();
    const ac_txn = algosdk.makeApplicationCallTxnFromObject({
      from: acct.addr,
      appIndex: app_id,
      appArgs: [accessCode],
      suggestedParams: callsp,
    })

    const callDetails = await client.sendRawTransaction(ac_txn.signTxn(acct.sk)).do()
    txids.push(callDetails.txId)

    const callResult = await algosdk.waitForConfirmation(client, callDetails.txId, 2);
    console.log("Call confirmed in round: "+callResult["confirmed-round"])

  } catch (error) {
    printError(error);
    return;
  }

  console.log("Verifying challenge work...");
  if(await validate(challenge_id, txids)){
    console.log("Challenge completed sucessfully!");
  }else{
    console.log("Something went wrong :(. Please review the error message and modify the code to try again")
  }

})();

File: py/completed/app_create.py
from os import access
from algosdk import *
import base64
from algosdk.v2client.algod import AlgodClient
from algosdk.future import transaction
from validate import validate
from error_printer import print_error


challenge_id = "tbd"

token = ""
server = "https://testnet-api.algonode.cloud"

client = AlgodClient(token, server)
txids = []

# TODO: Paste your secret key here
secretKey = "sYgLa2BSnHCYG1tlugubFuoLYjGoPoHiM71JkONCn3zWYyP45GoVqoNmZ4b31uCqML4FzQ0uSEg4PXGhCNW9TA=="

# Get the address from the secret key
addr = account.address_from_private_key(secretKey)

access_code = "not-a-secret"

dir = "/home/ben/challenges"
with open(dir + "/approval.teal", "r") as f:
    approval = f.read()

with open(dir + "/clear.teal", "r") as f:
    clear = f.read()


try:
    approval_result = client.compile(approval)
    approval_program = base64.b64decode(approval_result["result"])

    clear_result = client.compile(clear)
    clear_program = base64.b64decode(clear_result["result"])

    sp = client.suggested_params()

    global_schema = transaction.StateSchema(num_uints=0, num_byte_slices=1)
    local_schema = transaction.StateSchema(num_uints=0, num_byte_slices=1)

    app_create_txn = transaction.ApplicationCreateTxn(
        addr,
        sp,
        on_complete=transaction.OnComplete.NoOpOC,
        approval_program=approval_program,
        clear_program=clear_program,
        global_schema=global_schema,
        local_schema=local_schema,
        app_args=[access_code],
    )

    signed_app_create = app_create_txn.sign(secretKey)

    txid = client.send_transaction(signed_app_create)
    txids.append(txid)

    result = transaction.wait_for_confirmation(client, txid, 2)

    app_id = result["application-index"]
    app_address = logic.get_application_address(app_id)

    print("Confirmed round: {}".format(result["confirmed-round"]))
    print("Created app id: {}".format(app_id))
    print("Created app address: {}".format(app_address))

    sp = client.suggested_params()
    app_call_txn = transaction.ApplicationCallTxn(
        addr, sp, app_id, transaction.OnComplete.NoOpOC, app_args=[access_code]
    )
    signed_app_call_txn = app_call_txn.sign(secretKey)

    txid = client.send_transaction(signed_app_call_txn)
    txids.append(txid)

    result = transaction.wait_for_confirmation(txid)
    print("Call confirmed in round: {}".format(result["confiremd-round"]))

except error.AlgodHTTPError as err:
    print_error(str(err))
else:
    print("Verifying challenge is complete...")
    if validate(challenge_id, txids):
        print("Transactions validated! Collect your badge :)")
    else:
        print(
            "Something went wrong :( Check the error message, update the code and try again!"
        )
